File: scripts/check-i18n-keys.js
/**
 * Script để rà soát toàn bộ code và tìm các key i18n còn thiếu
 * Chạy: node scripts/check-i18n-keys.js
 */

const fs = require('fs');
const path = require('path');

// Đường dẫn thư mục gốc
const ROOT_DIR = path.join(__dirname, '..');
const LANG_DIR = path.join(ROOT_DIR, 'lang');

// Đọc các file ngôn ngữ
function loadLangFiles() {
  const langFiles = {};
  const files = fs.readdirSync(LANG_DIR).filter(f => f.endsWith('.json'));
  
  for (const file of files) {
    const langCode = file.replace('.json', '');
    const filePath = path.join(LANG_DIR, file);
    try {
      const content = fs.readFileSync(filePath, 'utf8');
      langFiles[langCode] = JSON.parse(content);
      console.log(`✓ Đã tải ${file} (${Object.keys(langFiles[langCode]).length} keys)`);
    } catch (error) {
      console.error(`✗ Lỗi khi đọc ${file}:`, error.message);
    }
  }
  
  return langFiles;
}

// Tìm tất cả các key được sử dụng trong code
function findUsedKeys(dir, extensions = ['.js', '.html']) {
  const usedKeys = new Set();
  const patterns = [
    // window.Lang.get('key')
    /window\.Lang\.get\(['"]([^'"]+)['"]/g,
    // Lang.get('key')
    /Lang\.get\(['"]([^'"]+)['"]/g,
    // getLang('key')
    /getLang\(['"]([^'"]+)['"]/g,
    // data-i18n="key"
    /data-i18n=['"]([^'"]+)['"]/g,
    // data-i18n-placeholder="key"
    /data-i18n-placeholder=['"]([^'"]+)['"]/g,
    // data-i18n-title="key"
    /data-i18n-title=['"]([^'"]+)['"]/g,
    // data-i18n-value="key"
    /data-i18n-value=['"]([^'"]+)['"]/g,
  ];

  function scanFile(filePath) {
    try {
      const content = fs.readFileSync(filePath, 'utf8');
      
      for (const pattern of patterns) {
        let match;
        while ((match = pattern.exec(content)) !== null) {
          usedKeys.add(match[1]);
        }
      }
    } catch (error) {
      // Bỏ qua lỗi đọc file
    }
  }

  function scanDirectory(dirPath) {
    try {
      const entries = fs.readdirSync(dirPath, { withFileTypes: true });
      
      for (const entry of entries) {
        const fullPath = path.join(dirPath, entry.name);
        
        // Bỏ qua node_modules, .git, và các thư mục không cần thiết
        if (entry.name.startsWith('.') || 
            entry.name === 'node_modules' || 
            entry.name === 'scripts' ||
            entry.name === 'lang') {
          continue;
        }
        
        if (entry.isDirectory()) {
          scanDirectory(fullPath);
        } else if (entry.isFile()) {
          const ext = path.extname(entry.name);
          if (extensions.includes(ext)) {
            scanFile(fullPath);
          }
        }
      }
    } catch (error) {
      // Bỏ qua lỗi đọc thư mục
    }
  }

  scanDirectory(dir);
  return usedKeys;
}

// So sánh keys
function compareKeys(usedKeys, langFiles) {
  const results = {};
  
  for (const [langCode, langKeys] of Object.entries(langFiles)) {
    const missing = [];
    const extra = [];
    
    // Tìm keys thiếu
    for (const usedKey of usedKeys) {
      if (!(usedKey in langKeys)) {
        missing.push(usedKey);
      }
    }
    
    // Tìm keys thừa (có trong file nhưng không dùng)
    for (const langKey of Object.keys(langKeys)) {
      if (!usedKeys.has(langKey)) {
        extra.push(langKey);
      }
    }
    
    results[langCode] = {
      missing: missing.sort(),
      extra: extra.sort(),
      total: Object.keys(langKeys).length,
      used: usedKeys.size
    };
  }
  
  return results;
}

// Main
console.log('🔍 Đang rà soát i18n keys...\n');

// 1. Tải các file ngôn ngữ
console.log('📂 Đang tải các file ngôn ngữ...');
const langFiles = loadLangFiles();
console.log('');

// 2. Tìm tất cả keys được sử dụng
console.log('🔎 Đang quét code để tìm các key được sử dụng...');
const usedKeys = findUsedKeys(ROOT_DIR);
console.log(`✓ Tìm thấy ${usedKeys.size} keys được sử dụng trong code\n`);

// 3. So sánh
console.log('📊 Đang so sánh...\n');
const results = compareKeys(usedKeys, langFiles);

// 4. Hiển thị kết quả
for (const [langCode, result] of Object.entries(results)) {
  console.log(`\n${'='.repeat(60)}`);
  console.log(`📄 ${langCode}.json`);
  console.log(`${'='.repeat(60)}`);
  console.log(`Tổng số keys: ${result.total}`);
  console.log(`Keys được sử dụng: ${result.used}`);
  console.log(`Keys thiếu: ${result.missing.length}`);
  console.log(`Keys thừa (không dùng): ${result.extra.length}`);
  
  if (result.missing.length > 0) {
    console.log(`\n❌ CÁC KEY THIẾU (${result.missing.length}):`);
    result.missing.forEach(key => {
      console.log(`   - "${key}"`);
    });
  }
  
  if (result.extra.length > 0 && result.extra.length < 50) {
    console.log(`\n⚠️  CÁC KEY THỪA (${result.extra.length} - có thể không dùng):`);
    result.extra.slice(0, 20).forEach(key => {
      console.log(`   - "${key}"`);
    });
    if (result.extra.length > 20) {
      console.log(`   ... và ${result.extra.length - 20} keys khác`);
    }
  }
}

// 5. Tạo file báo cáo JSON
const reportPath = path.join(ROOT_DIR, 'i18n-report.json');
const report = {
  timestamp: new Date().toISOString(),
  usedKeysCount: usedKeys.size,
  results: results
};

fs.writeFileSync(reportPath, JSON.stringify(report, null, 2));
console.log(`\n\n📝 Đã tạo báo cáo chi tiết tại: ${reportPath}`);

// 6. Tổng kết
const allMissing = new Set();
for (const result of Object.values(results)) {
  result.missing.forEach(key => allMissing.add(key));
}

if (allMissing.size > 0) {
  console.log(`\n\n⚠️  TỔNG KẾT: Còn ${allMissing.size} keys thiếu trong ít nhất một file ngôn ngữ`);
  process.exit(1);
} else {
  console.log(`\n\n✅ TẤT CẢ CÁC KEY ĐỀU ĐÃ CÓ ĐẦY ĐỦ!`);
  process.exit(0);
}

File: scripts/check-i18n-keys.py
#!/usr/bin/env python3
"""
Script để rà soát toàn bộ code và tìm các key i18n còn thiếu
Chạy: python scripts/check-i18n-keys.py
"""

import os
import json
import re
from pathlib import Path

ROOT_DIR = Path(__file__).parent.parent
LANG_DIR = ROOT_DIR / 'lang'

# Patterns để tìm keys
PATTERNS = [
    r"window\.Lang\.get\(['\"]([^'\"]+)['\"]",
    r"Lang\.get\(['\"]([^'\"]+)['\"]",
    r"getLang\(['\"]([^'\"]+)['\"]",
    r"data-i18n=['\"]([^'\"]+)['\"]",
    r"data-i18n-placeholder=['\"]([^'\"]+)['\"]",
    r"data-i18n-title=['\"]([^'\"]+)['\"]",
    r"data-i18n-value=['\"]([^'\"]+)['\"]",
]

def load_lang_files():
    """Tải tất cả các file ngôn ngữ"""
    lang_files = {}
    for file_path in LANG_DIR.glob('*.json'):
        lang_code = file_path.stem
        try:
            with open(file_path, 'r', encoding='utf-8') as f:
                lang_files[lang_code] = json.load(f)
            print(f"✓ Đã tải {file_path.name} ({len(lang_files[lang_code])} keys)")
        except Exception as e:
            print(f"✗ Lỗi khi đọc {file_path.name}: {e}")
    return lang_files

def find_used_keys(root_dir):
    """Tìm tất cả các key được sử dụng trong code"""
    used_keys = set()
    
    # Các thư mục và file cần bỏ qua
    ignore_dirs = {'.git', 'node_modules', 'scripts', 'lang', '__pycache__'}
    ignore_files = {'.gitignore', '.gitattributes'}
    
    # Các extension cần quét
    extensions = {'.js', '.html', '.jsx', '.ts', '.tsx'}
    
    def scan_file(file_path):
        try:
            with open(file_path, 'r', encoding='utf-8', errors='ignore') as f:
                content = f.read()
                for pattern in PATTERNS:
                    matches = re.findall(pattern, content)
                    used_keys.update(matches)
        except Exception:
            pass
    
    def scan_directory(dir_path):
        try:
            for item in dir_path.iterdir():
                if item.name.startswith('.') or item.name in ignore_dirs:
                    continue
                if item.is_dir():
                    scan_directory(item)
                elif item.is_file() and item.suffix in extensions:
                    if item.name not in ignore_files:
                        scan_file(item)
        except Exception:
            pass
    
    scan_directory(root_dir)
    return used_keys

def compare_keys(used_keys, lang_files):
    """So sánh keys được sử dụng với keys có trong file ngôn ngữ"""
    results = {}
    
    for lang_code, lang_keys in lang_files.items():
        missing = sorted([k for k in used_keys if k not in lang_keys])
        extra = sorted([k for k in lang_keys if k not in used_keys])
        
        results[lang_code] = {
            'missing': missing,
            'extra': extra,
            'total': len(lang_keys),
            'used': len(used_keys)
        }
    
    return results

def main():
    print('🔍 Đang rà soát i18n keys...\n')
    
    # 1. Tải các file ngôn ngữ
    print('📂 Đang tải các file ngôn ngữ...')
    lang_files = load_lang_files()
    print()
    
    # 2. Tìm tất cả keys được sử dụng
    print('🔎 Đang quét code để tìm các key được sử dụng...')
    used_keys = find_used_keys(ROOT_DIR)
    print(f"✓ Tìm thấy {len(used_keys)} keys được sử dụng trong code\n")
    
    # 3. So sánh
    print('📊 Đang so sánh...\n')
    results = compare_keys(used_keys, lang_files)
    
    # 4. Hiển thị kết quả
    all_missing = set()
    for lang_code, result in results.items():
        print(f"\n{'='*60}")
        print(f"📄 {lang_code}.json")
        print(f"{'='*60}")
        print(f"Tổng số keys: {result['total']}")
        print(f"Keys được sử dụng: {result['used']}")
        print(f"Keys thiếu: {len(result['missing'])}")
        print(f"Keys thừa (không dùng): {len(result['extra'])}")
        
        if result['missing']:
            print(f"\n❌ CÁC KEY THIẾU ({len(result['missing'])}):")
            for key in result['missing']:
                print(f"   - \"{key}\"")
                all_missing.add(key)
        
        if result['extra'] and len(result['extra']) < 50:
            print(f"\n⚠️  CÁC KEY THỪA (mẫu - {len(result['extra'])} tổng cộng):")
            for key in result['extra'][:20]:
                print(f"   - \"{key}\"")
            if len(result['extra']) > 20:
                print(f"   ... và {len(result['extra']) - 20} keys khác")
    
    # 5. Tổng kết
    print(f"\n\n{'='*60}")
    if all_missing:
        print(f"⚠️  TỔNG KẾT: Còn {len(all_missing)} keys thiếu trong ít nhất một file ngôn ngữ")
        print("\nDanh sách keys cần thêm:")
        for key in sorted(all_missing):
            print(f"  - {key}")
        return 1
    else:
        print("✅ TẤT CẢ CÁC KEY ĐỀU ĐÃ CÓ ĐẦY ĐỦ!")
        return 0

if __name__ == '__main__':
    exit(main())

